File: custom_finance/public/js/fees.js
frappe.ui.form.on('Fees', {
    // setup(frm){
    //     frm.set_query("fee_structure", function() {
    //         return {
    //             query: 'ed_tec.ed_tec.doctype.fees.get_fee_structures',
    //             filters: {
    //                 "exam_application":frm.doc.exam_application
    //             }
    //         };
    //     });
    // },

    /////////////// My Code for Filter of table/////
    onload: function(frm) {
		frm.set_query("receivable_account","components", function(_doc, cdt, cdn) {
			var d = locals[cdt][cdn];
			return {
				filters: {
					'company': d.company,
					'account_type': d.account_type = 'Receivable',
					'is_group': d.is_group = 0
				}
			};
		});
		frm.set_query("income_account","components", function(_doc, cdt, cdn) {
			var d = locals[cdt][cdn];
			return {
				filters: {
					'company': d.company,
					'account_type': d.account_type = 'Income Account',
					'is_group': d.is_group = 0
				}
			};
		});
		erpnext.accounts.dimensions.setup_dimension_filters(frm, frm.doctype);
	},
	//////////////////////////// end of my Code //////////////////////////
    refresh(frm){
        if(frm.doc.docstatus===1 && frm.doc.outstanding_amount==0) {
			frm.add_custom_button(__("Return/Refund"), function() {
                frappe.model.open_mapped_doc({
					method: "ed_tec.ed_tec.doctype.fees.make_refund_fees",
					frm: frm,
				});
			});
		}
    },
    amount(frm){
            if(frm.doc.amount){
                // frm.doc.outstanding_amount = frm.doc.grand_total - frm.doc.amount;
                // frm.doc.waiver_amount = frm.doc.amount
                frm.set_value("outstanding_amount",frm.doc.grand_total - frm.doc.amount);
                frm.set_value("waiver_amount",frm.doc.amount);

            }
     },
    percentage(frm){
        if(frm.doc.percentage){
            // outstanding_amount = frm.doc.grand_total - (frm.doc.grand_total*(frm.doc.percentage/100));
            // waiver_amount = (frm.doc.grand_total*(frm.doc.percentage/100))
            frm.set_value("outstanding_amount",frm.doc.grand_total - (frm.doc.grand_total*(frm.doc.percentage/100)));
            frm.set_value("waiver_amount",(frm.doc.grand_total*(frm.doc.percentage/100)));
        }
    },
    student(frm){
        if (frm.doc.student){
            frm.trigger("set_program_enrollment");
            frm.set_query("programs", function() {
                return {
                    query: 'ed_tec.ed_tec.doctype.fees.get_progarms',
                    filters: {
                        "student":frm.doc.student
                    }
                };
            });
            frm.set_query("program", function() {
                return {
                    query: 'ed_tec.ed_tec.doctype.fees.get_sem',
                    filters: {
                        "student":frm.doc.student
                    }
                };
            });
            frm.set_query("academic_term", function() {
                return {
                    query: 'ed_tec.ed_tec.doctype.fees.get_term',
                    filters: {
                        "student":frm.doc.student
                    }
                };
            });
            frm.set_query("academic_year", function() {
                return {
                    query: 'ed_tec.ed_tec.doctype.fees.get_year',
                    filters: {
                        "student":frm.doc.student
                    }
                };
            });
            frm.set_query("student_category", function() {
                return {
                    query: 'ed_tec.ed_tec.doctype.fees.get_student_category',
                    filters: {
                        "student":frm.doc.student
                    }
                };
            });
            frm.set_query("student_batch", function() {
                return {
                    query: 'ed_tec.ed_tec.doctype.fees.get_batch',
                    filters: {
                        "student":frm.doc.student
                    }
                };
            });
            frm.set_query("fee_structure", function() {
                return {
                    query: 'ed_tec.ed_tec.doctype.fees.get_fee_structures',
                    filters: {
                        "student":frm.doc.student
                    }
                };
            });
            
        }
    },
	set_program_enrollment(frm) {
        frappe.call({
            method: "ed_tec.ed_tec.doctype.program_enrollment.get_program_enrollment",
            args: {
                student: frm.doc.student,
            },
            callback: function(r) { 
                if (r.message){
                    frm.set_value("program_enrollment",r.message['name'])
                }
            } 
            
        });    
        
	},
    setup(frm){
        frm.set_query("fees_category","components", function() {
            return {
                query: 'ed_tec.ed_tec.doctype.fees.get_fees_category',
                filters: {
                    "fee_structure":frm.doc.fee_structure
                }
            };
        });

    }
})
frappe.ui.form.on("Fee Component", "amount", function(frm, cdt, cdn) {
    var d = locals[cdt][cdn];

    if(d.amount){
        d.waiver_on_amount = d.amount 
        refresh_field("waiver_on_amount", d.name, d.parentfield);
    }
})
frappe.ui.form.on("Fee Component", "waiver_amount", function(frm, cdt, cdn) {
    var d = locals[cdt][cdn];
    if(d.waiver_amount && d.amount){
        d.amount =  d.amount -d.waiver_amount
       d.total_waiver_amount  = d.waiver_amount
        refresh_field("amount", d.name, d.parentfield);
        refresh_field("total_waiver_amount", d.name, d.parentfield);
  
    }
    if(!d.amount){
        frappe.throw("Please add Amount first");
    }
});


frappe.ui.form.on("Fee Component", "percentage", function(frm, cdt, cdn) {
    var d = locals[cdt][cdn];
    if(d.percentage && d.amount){
        d.amount =  d.amount - ((d.percentage/100) * d.amount)
       d.total_waiver_amount  = ((d.percentage/100) * d.amount)
        refresh_field("amount", d.name, d.parentfield);
        refresh_field("total_waiver_amount", d.name, d.parentfield);
  
    }
    if(!d.amount){
        frappe.throw("Please add Amount first");
    }
});


/////////////// My Code for Filter of table/////
frappe.ui.form.on("Fee Component", "amount", function(frm, cdt, cdn) {
   
    var ed_details = frm.doc.components;
    for(var i in ed_details) {
            
    if (ed_details[i].amount) {
        // ed_details[i].total_fee_amount="15";
        ed_details[i].grand_fee_amount=ed_details[i].amount;
    } 
   }
        cur_frm.refresh_field ("components");
    
});
frappe.ui.form.on("Fee Component", "amount", function(frm, cdt, cdn) {

    var ed_details = frm.doc.components;
    for(var i in ed_details) {
            
    if (ed_details[i].amount) {
        // ed_details[i].total_fee_amount="15";
        ed_details[i].outstanding_fees=ed_details[i].amount;
    }	 
        
}
    cur_frm.refresh_field ("components");

});
////// End of my code//////

File: custom_finance/custom_finance/validations/fees.py
import frappe
from frappe.model.mapper import get_mapped_doc
from kp_edtec.kp_edtec.utils import duplicate_row_validation

def on_submit(doc,method):
    if doc.exam_application:
        ex=frappe.get_doc("Exam Application",doc.exam_application)
        ex.status="Paid"
        ex.flags._validate_update_after_submit = True
        ex.submit()
        
    if doc.program_enrollment and not doc.programs:
        doc.programs=frappe.db.get_value("Program Enrollment",doc.program_enrollment,'programs')
    
    if doc.is_return and doc.return_against:
        return_against=frappe.get_doc("Fees",doc.return_against)
        return_against.return_issued=1
        return_against.submit()

def on_cancel(doc,method):
     if doc.is_return and doc.return_against:
        return_against=frappe.get_doc("Fees",doc.return_against)
        return_against.return_issued=0
        return_against.submit()

def validate(doc,method):
    validate_amount(doc)
    duplicate_row_validation(doc, "components", ['fees_category', 'amount'])
    # doc.calculate_amount()

# def calculate_amount(self):
# 		for events in self.get("components"):
# 			events.grand_fee_amount=events.amount
# 			events.outstanding_fees=events.amount

def validate_amount(doc):
    if doc.is_return:
        for cmp in doc.components:
            if cmp.amount>0:
                frappe.throw("Component <b>{0}</b> amount Must be <b>-ve</b> value".format(cmp.fees_category))
        if doc.grand_total>0:
            frappe.throw("Grand Total Must be <b>-ve</b> value")

@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def get_fee_structures(doctype, txt, searchfield, start, page_len, filters):
    program=""
    for d in frappe.get_all("Current Educational Details",{"parent":filters.get("student")},['semesters']):
        program+=d.semesters
    return frappe.db.sql("""select name,program,student_category,academic_year from `tabFee Structure` where program IN ('{0}') and (name like '%{1}%' or program like '%{1}%' or student_category like '%{1}%' or academic_year like '%{1}%')""".format(program,txt))

@frappe.whitelist()
def get_progarms(doctype, txt, searchfield, start, page_len, filters):
    fltr = {'docstatus':1}
    if txt:
        fltr.update({"programs":txt})

    fltr.update({"student":filters.get("student")})
    data = frappe.get_all("Program Enrollment",fltr,['programs'],as_list=1)
    return data

@frappe.whitelist()
def get_sem(doctype, txt, searchfield, start, page_len, filters):
    fltr = {'docstatus':1}
    if txt:
        fltr.update({"program":txt})

    fltr.update({"student":filters.get("student")})
    data = frappe.get_all("Program Enrollment",fltr,['program'],as_list=1)
    return data 


@frappe.whitelist()
def get_term(doctype, txt, searchfield, start, page_len, filters):
    fltr = {'docstatus':1}
    if txt:
        fltr.update({"academic_term":txt})

    fltr.update({"student":filters.get("student")})
    data = frappe.get_all("Program Enrollment",fltr,['academic_term'],as_list=1)
    return data

@frappe.whitelist()
def get_year(doctype, txt, searchfield, start, page_len, filters):
    fltr = {'docstatus':1}
    if txt:
        fltr.update({"academic_year":txt})

    fltr.update({"student":filters.get("student")})
    data = frappe.get_all("Program Enrollment",fltr,['academic_year'],as_list=1)
    return data 

@frappe.whitelist()
def get_batch(doctype, txt, searchfield, start, page_len, filters):
    fltr = {'docstatus':1}
    if txt:
        fltr.update({"student_batch_name":txt})

    fltr.update({"student":filters.get("student")})
    data = frappe.get_all("Program Enrollment",fltr,['student_batch_name'],as_list=1)
    return data 


@frappe.whitelist()
def get_student_category(doctype, txt, searchfield, start, page_len, filters):
    fltr = {}
    lst = []
    if txt:
        fltr.update({"student_category":txt})
    fltr.update({"student":filters.get("student")})
    for i in frappe.get_all("Program Enrollment",fltr,['student_category']):
        if i.student_category not in lst:
            lst.append(i.student_category)
    return [(d,) for d in lst]

@frappe.whitelist()
def get_fees_category(doctype, txt, searchfield, start, page_len, filters):
    fltr = {}
    lst = []
    if txt:
        fltr.update({"fees_category":txt})
    
    fltr.update({"parent":filters.get("fee_structure"),"parenttype":"Fee Structure"})
    for i in frappe.get_all("Fee Component",fltr,['fees_category'],order_by="idx asc"):
        if i.fees_category not in lst:
            lst.append(i.fees_category)
    return [(d,) for d in lst]

@frappe.whitelist()
def make_refund_fees(source_name, target_doc=None):
    def set_missing_values(source, target):
        target.set("components",[])
        for d in source.get("components"):
            target.append("components",{
					"fees_category":d.fees_category,
					"amount":-d.amount,
					"description":d.description,
                    "income_account":d.income_account,
                    "receivable_account":d.receivable_account
				})
        target.grand_total=(-source.grand_total)
        target.is_return=1
        target.return_against=source.name
        target.outstanding_amount=(-source.outstanding_amount)

    doclist = get_mapped_doc("Fees", source_name, 	{
        "Fees": {
            "doctype": "Fees",
        },
    }, target_doc, set_missing_values)

    return doclist
@frappe.whitelist()
def get_fee_components(fee_structure):
    """Returns Fee Components.

    :param fee_structure: Fee Structure.
    """
    if fee_structure:
        fs = frappe.get_all("Fee Component", fields=["fees_category", "description", "amount","receivable_account",
                                                    "income_account","grand_fee_amount","outstanding_fees","waiver_type","percentage","waiver_amount",
                                                    "total_waiver_amount"] , 
                                                filters={"parent": fee_structure}, order_by= "idx asc")
        return fs

@frappe.whitelist()
def get_program_enrollment(student):
    data=frappe.get_all("Program Enrollment",{'student':student,'docstatus':1},['name','program','programs'],order_by= "name desc")
    if len(data)>0:
        return data[0]

